src/main.rs

```diff
@@ -373,31 +376,46 @@ fn run(args: Cli) -> Result<i32> {
         ),
     };
     let pretty = args.pretty.unwrap_or_else(|| buffer.guess_pretty());
-    let mut printer = Printer::new(print.clone(), pretty, args.style, args.stream, buffer);
+    let mut printer = Printer::new(pretty, args.style, args.stream, buffer);
 
-    printer.print_request_headers(&request, &*cookie_jar)?;
-    printer.print_request_body(&mut request)?;
+    if print.request_headers {
+        printer.print_request_headers(&request, &*cookie_jar)?;
+    }
+    if print.request_body {
+        printer.print_request_body(&mut request)?;
+    }
 
     if !args.offline {
-        let response = if args.follow {
-            let mut client =
-                redirect::RedirectFollower::new(&client, args.max_redirects.unwrap_or(10));
-            if let Some(history_print) = args.history_print {
-                printer.print = history_print;
+        let response = {
+            let history_print = args.history_print.unwrap_or(print);
+            let mut client = ClientWithMiddleware::new(&client);
+            if args.follow {
+                let mut redirect_follower = RedirectFollower::new(args.max_redirects.unwrap_or(10));
+                if args.all {
+                    redirect_follower.on_redirect(|prev_response, next_request| {
+                        if history_print.response_headers {
+                            printer.print_response_headers(&prev_response)?;
+                        }
+                        if history_print.response_body {
+                            printer.print_response_body(prev_response)?;
+                            printer.print_separator()?;
+                        }
+                        if history_print.request_headers {
+                            printer.print_request_headers(next_request, &*cookie_jar)?;
+                        }
+                        if history_print.request_body {
+                            printer.print_request_body(next_request)?;
+                        }
+                        Ok(())
+                    });
+                }
+                client = client.with(redirect_follower);
             }
-            if args.all {
-                client.on_redirect(|prev_response, next_request| {
-                    printer.print_response_headers(&prev_response)?;
-                    printer.print_response_body(prev_response)?;
-                    printer.print_separator()?;
-                    printer.print_request_headers(next_request, &*cookie_jar)?;
-                    printer.print_request_body(next_request)?;
-                    Ok(())
-                });
+            if let Some(Auth::Digest(username, password)) = &auth {
+                // TODO: print intermediary requests/responses if --all flag is used
```

Comment:
This is a little bit tricky since `Printer` cannot be easily shared between multiple middlewares. I have thought about cloning `Printer` but that would mean I will have to clone `Buffer` which can be a file handle.

Reply:
I have opted to store the printer callback in `ClientWithMiddleware` instead